### PR TITLE
feat: improve component based tests approach

### LINKED ISSUE #5922: RFC: Improve component based tests approach
### Current behavior:
The current state of cypress "unit tests" (they are actually end-to-end tests if you are building components library/design systems/everything) is completely unusable.
For example library https://github.com/bahmutov/cypress-react-unit-test is not allowing to tests portals approach,  impossible to test CSS-in-js, and anything that needs access to the document object, approach. 

### Reason
The main problem: when we want to render the application from the tests (like we have `cy.mount`) we understand that our tests are running from the different iframe. And this is the main problem for the component-based tests approach. 

### Request for change:
I propose to add a new config field like `mountMode: true` and just load an application right to the `.aut-frame`. This will fix all the problems 
> P.S. Already have a working code example will make a PR right in a few mins
> P.P.S I believe that cypress tests for components can take over the world. Because this is much more reliable than jest, enzyme, testing-library or anything. Moreover, it is super interactive so it is possible to use it as a kind of replacement for storybook.

### LINKED ISSUE #6968: Show separate specs tree view for integration vs component specs
This is for https://github.com/cypress-io/cypress/pull/5923/ (component testing)

**against branch** `issue-5922-mountMode`

Example project with integration and component tests https://github.com/bahmutov/test-mdx-example (clone and install dependencies)

To run the example project, switch Cypress monorepo to `issue-5922-mountMode`, run `yarn` to install dependencies and build, and then use

```
yarn start --project <path to test-md-example folder>
```

If I print specs received by the Desktop GUI `specs.integration` like this
```js
import ipc from '../lib/ipc'
import specsStore from './specs-store'

const getSpecs = (setProjectError) => {
  const DEFAULT_SPECS = {
    integration: [],
  }

  ipc.getSpecs((err, specs = DEFAULT_SPECS) => {
    if (err) {
      return setProjectError(err)
    }

    console.log('specs')
    console.table(specs.integration)

    specsStore.setSpecs(specs)
  })
}
```

The list shows `specType` property

Screen Shot 2020-04-08 at 3 11 20 PM

I think we should separate integration specs from component specs, but could not figure out how to do this in specs jsx file. Probably something like this would be nice

Screen Shot 2020-04-08 at 3 11 20 PM

### PR DESCRIPTION
**Pull request by dmtrKovalenko**
- Closes #5922
- Closes #6968

### User facing changelog

```json
{
  "experimentalComponentTesting": true,
  "componentFolder": "src"
}
```

`componentFolder` has default value in `config.coffee`.

### Additional details

The current way of copying a component registered in spec iframe does not work very well (or at all if using React hooks). This change uses the spec iframe as its only iframe and fixes all issues. I could not find a case except for some edge cases where it did not do the right thing. See https://github.com/bahmutov/cypress-react-unit-test/tree/feature/cypress-mount-mode#large-examples and https://github.com/topics/cypress-react-unit-test-example and 100 specs in https://github.com/bahmutov/cypress-react-unit-test/pull/108

But the same applies to all frameworks, and I have made an open PR to cypress-vue-unit-test as well https://github.com/bahmutov/cypress-vue-unit-test/pull/247

### How has the user experience changed?
For current cypress users - no way

For applications that use React Hooks for example - a new feature that allows replacing all Jest browser emulation unit tests with true browser tests, see the [cypress-react-unit-test README](https://github.com/bahmutov/cypress-react-unit-test/tree/feature/cypress-mount-mode)

### PR Tasks

- [x] merge latest `develop` branch
- [x] make sure `Run all specs` button works (should only run `integration` specs)
- [x] watch both integration folder and component test folder
- [ ] ⚠️ we apply the same `testFiles` pattern in both folders, which might not work too well. Workaround - use different config JSON files.
- [x] update Desktop GUI to show `integration` tests separately from `component` tests, probably different top level heading #6968
- [x] Have tests been added/updated?
- [x] Has the original issue been tagged with a release in ZenHub? 
- [x] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? https://github.com/cypress-io/cypress-documentation/pull/2745
- [x] Have API changes been updated in the [`type definitions`](cli/types/index.d.ts)?
- [x] Have new configuration options been added to the [`cypress.schema.json`](cli/schema/cypress.schema.json)?

### CODE REVIEW

**Comment by dmtrKovalenko** [comment]
I am not sure about where to write tests for this change, didn't find appropriate test files. Please help :)

**Comment by jennifer-shehane** [comment]
@bahmutov What is blocking review of this PR?

**Comment by dmtrKovalenko** [comment]
Ok, question - can I continue work on that (enable it via file-pattern, and restricting e2e commands) in order to allow this functionality ASAP, or you are not interested in such global changes, or smth?

> Btw cy.visit continue work for some reason, we discussed that with Gleb.

**Comment by bahmutov** [comment]
I think we should get this working right, then iterate on how it would interact with E2E tests. For now, I am ok assuming that any spec mounting a component is NOT going to call `cy.visit`

**Comment by dmtrKovalenko** [comment]
We can basically throw an error when calling cy.visit in mount mode.

**Comment by brian-mann** [comment]
There are a lot more considerations here - beyond just configuration you also have things like lifecycle events. 

It doesn't necessarily make sense to do things like clear cookies, local storage, etc by default. Maybe those could be opt-in. You also would need to make decisions regarding how to "blow away" or clear the AUT between each test. Maybe you just need to wipe the DOM - or expose a seam to let the user choose.

There is potentially a significant number of new API's or surface area to expand on based on what is the typical use case. Like I said, this is a good start but there's going to be a lot more before we can officially support this. I'm wary of adding any kind of configuration option - especially one we know is immediately going to be deprecated. I'd rather go down the path of creating either beta flags or non-documented options.

I would prefer giving a seam inside of the runner code for you to write anything you want arbitrarily - but if you write that code in the `spec` itself it won't work because by then it would be too late. So that's not really an option unfortunately.

**Comment by brian-mann** [comment]
Another note - unit testing likely doesn't need the specs to be served independently from one another. You likely just want to run "all of them" together in CI as opposed to having them chunked apart - which we have effectively hard coded.

**Comment by brian-mann** [comment]
I have an entire proposal somewhere I should dig up...

**Comment by sainthkh** [comment]
@brian-mann Isn't #318 what you're looking for?

**Comment by bahmutov** [comment]
After playing with cypress-react-unit-test https://github.com/bahmutov/cypress-react-unit-test/pull/108 A LOT and going through examples shown by the React tutorials and official guides I have to say I ❤️ this component-testing experience.

But I would not add the `mountMode ` configuration option - since I see the same project writing both E2E and component tests. Just like @brian-mann said above, I would create parallel 2 settings to tell the Test Runner which files use which method

```
integrationFolder: "cypress/integration",
testFiles: "**/*.js"
```
new options
```
componentFolder: "cypress/component"
componentTestFiles: "**/*.js"
```

With the above changes, one useful thing I think we could support is the component tests right alongside the source files. For example, the project structure could look like this:

```
cypress/
  integration/
    app specs
src/
  App.js
  App.spec.js
  components/
    Header.js
    Header.spec.js
```

and the `cypress.json` file would say

```json
componentFolder: "src",
componentTestFiles: "**/*.spec.js"
```

**Comment by dmtrKovalenko** [comment]
@bahmutov. I could try to update PR with the proposed API. Should I try to do this?

**Comment by bahmutov** [comment]
@dmtrKovalenko yeah, if you can remove `mount` option and instead specify mount mode via component tests files, that would be the best way to get this done quicker.

I think we can address how component tests interact with integration tests separately. For example, `cy.visit` should throw an error while in component test mounting mode, or what happens if the same spec file is both in integration and component tests sets - we can worry about these questions later.

**Comment by bahmutov** [comment]
@dmtrKovalenko I will work on component tests during the weekend if you haven't

**Comment by dmtrKovalenko** [comment]
@bahmutov I didn't do anything. Pretty buzy last days, so please do anything you want :)

**Comment by bahmutov** [comment]
- [x] added `componentFolder` option to `cypress.json`
- [x] implemented spec search to search `integrationFolder` and `componentFolder`. The returned spec objects have additional property `type` that can be either `integration` or `component`
- [ ] desktop GUI should group integration tests vs component tests, right now they are all placed together

Screen Shot 2020-01-05 at 2 02 25 PM

- [ ] desktop GUI should pass the `type` property with the spec when you click on it to open
- [ ] when opening a spec / launching browser, we should set the config `mountMode` depending on the spec type in `iframes.jsx` currently

```js
// packages/runner/src/iframe/iframes.jsx
if (this.props.config.mountMode) {
      // In mount mode we need to render something right from spec file
      // So load application tests to the aut frame
      return this._loadSpecInIframe($autIframe, specSrc)
      .then(() => $autIframe)
    }
```

**Comment by bahmutov** [comment]
spec type is passed to desktop gui

Screen Shot 2020-01-05 at 2 18 23 PM

**Comment by bahmutov** [comment]
close - for component tests it seems to confuse it with "run all" mode because it is not an integration!

**Comment by bahmutov** [comment]
ok, changed the spec bundle to take into account integration vs component type when deciding the filename, now can click on spec in the desktop gui and correctly test it as an integration or component test

**Comment by bahmutov** [comment]
## TODO

- [ ] remove `mountMode` from `cypress.json` - no longer necessary, has been replaced by `componentFolder` setting
- [ ] add `componentFolder` setting to `cypress.json` schema

File: packages/server/lib/open_project.js [hunk 1]
```diff
@@ -189,13 +192,25 @@ const moduleFactory = () => {
...
+              debug(
+                'found \'%d\' specs using spec pattern \'%s\': %o',
+                names.length,
+                cfg.testFiles,
+                names
+              )
+            }
+
+            return {
+              integration: specs,
```
Comment (bahmutov) [inline packages/server/lib/open_project.js]: so right now we put all specs under `integration` property. Now the specs have `specType` property, which could be either `integration` or `component` - we should rename `integration` into `specs`

**Comment by bahmutov** [comment]
## TODO

- [ ] the way we determine how to serve the spec file in `server/lib/util/path_helpers.js` via `isComponentTestRe ` is a hack. Instead, the spec iframe should pass a parameter for the spec type when asking for the bundled spec file. Right now the only parameter is the cache-busting `t=...` URL parameter

**Comment by bahmutov** [comment]
## TODO

- [ ] show component spec files separately from the integration specs in desktop gui

Screen Shot 2020-01-05 at 2 18 23 PM

**Comment by jennifer-shehane** [comment]
I've added `[WIP]` to the title to signify there are changes still required before a final review is ready.

File: cli/schema/cypress.schema.json [hunk 2]
```diff
@@ -210,6 +210,11 @@
       ],
       "default": "bundled",
       "description": "If set to 'system', Cypress will try to find a Node.js executable on your path to use when executing your plugins. Otherwise, Cypress will use the Node version bundled with Cypress."
+    },
+    "mountMode": {
```
Comment (bahmutov) [inline cli/schema/cypress.schema.json]: obsolete, I will remove this

**Comment by bahmutov** [comment]
# TODO
- [ ] update "Run all specs" logic to run all _integration_ tests, without component tests

otherwise getting so many errors

```
 cypress:server:plugins promise rejected for id inv32 ':' multi ./cypress/Users/gleb/git/cypress-vue-unit-test/cypress/component/basic/spec.js
Module not found: Error: Can't resolve '/Users/gleb/git/cypress-vue-unit-test/cypress/Users/gleb/git/cypress-vue-unit-test/cypress/component/basic/spec.js' in '/Users/gleb/git/cypress-vue-unit-test'
resolve '/Users/gleb/git/cypress-vue-unit-test/cypress/Users/gleb/git/cypress-vue-unit-test/cypress/component/basic/spec.js' in '/Users/gleb/git/cypress-vue-unit-test'
  using description file: /Users/gleb/git/cypress-vue-unit-test/package.json (relative path: .)
    Field 'browser' doesn't contain a valid alias configuration
```

**Comment by bahmutov** [comment]
Screen Shot 2020-01-30 at 10 54 07 AM

File: packages/server/test/integration/http_requests_spec.coffee [hunk 3]
```diff
@@ -43,6 +43,13 @@ removeWhitespace = (c) ->
   c = str.lines(c).join(" ")
   c
 
+trimString = (s) -> _.trim(s)
+nonEmpty = (s) -> !_.isEmpty(s)
+
+# removes leading whitespace and blank lines
+removeLeadingWhitespace = (text) ->
```
Comment (bahmutov) [inline packages/server/test/integration/http_requests_spec.coffee]: Before it was hard to see what is going one in a blob of text

Screen Shot 2020-03-17 at 2 09 17 PM

Now it is pretty clear
Screen Shot 2020-03-17 at 2 16 21 PM

**Comment by NickBolles** [comment]
@bahmutov what's the status of this? Is there anything I can help with? looks like a great idea and I'm using some workarounds in my current project that this would solve.

**Comment by bahmutov** [comment]
Need to figure out why custom components folder is not working.

```json
{
  "experimentalComponentTesting": true,
  "componentFolder": "src"
}
```

Integration test fetches from `cypress/integration` is working from default `cypress/integration`

```
GET /__cypress/iframes/integration/sanity.cy-spec.js 200
```

But fetching the component spec file does not work

```
GET /__cypress/src/counter/components/Counter.cy-spec.js 404
```

**Comment by bahmutov** [comment]
Seems there is relative path that breaks the aut source url
Screen Shot 2020-04-03 at 2 12 39 PM

File: packages/server/lib/open_project.js [hunk 4]
```diff
@@ -168,21 +174,39 @@ const moduleFactory = () => {
...
+            ignoreInitial: true,
+          })
+
+          this.specsWatcher.on('add', checkForSpecUpdates)
+
+          this.specsWatcher.on('unlink', checkForSpecUpdates)
         }
 
-        debug('watch test files: %s in %s', cfg.testFiles, cfg.integrationFolder)
+        if (experimentalComponentTestingEnabled && !this.componentSpecsWatcher) {
```
Comment (bahmutov) [inline packages/server/lib/open_project.js:197]: creating second watcher if component testing is enabled

**Comment by NickBolles** [comment]
@bahmutov Cypress noob here. Whats the best way to test this PR? I tried installing from the branch (`yarn add cypress#issue-5922-mountMode`) but that didn't seem to work.

File: cli/schema/cypress.schema.json [hunk 5]
```diff
@@ -103,6 +103,19 @@
       "default": "cypress/integration",
       "description": "Path to folder containing integration test files"
     },
+    "experimentalComponentTesting": {
+      "type": "boolean",
+      "default": false,
+      "description": "Enabled experimental component testing, see https://github.com/cypress-io/cypress/issues/5922"
+    },
+    "componentFolder": {
+      "type": [
```
Comment (NickBolles) [inline cli/schema/cypress.schema.json:112]: could this be an array? it seems like components are more likely than E2E tests to be in multiple folders.

**Comment by NickBolles** [comment]
For my component tests, I don't start up the server, I still see this warning. Maybe reworking this message to specify "integration tests will likely fail"?
image

**Comment by bahmutov** [comment]
@NicholasBoll Agree on the message, will need to decide later how to handle it. I have added tooltip to that button "Run all specs" to say "run all integration specs", but right now I don't see how it would work with multiple files, you cannot concatenate an integration and component specs files together for example. Also, we really hope to remove that button in the future, since running all specs is bad practice in our opinion.

**Comment by bahmutov** [comment]
Oopts, `#copyPackages linux` when building the binary does not understand `! wildcard`

```
packages/driver/package.json -> /root/cypress/dist/linux/packages/driver/package.json
packages/driver/patches -> /root/cypress/dist/linux/packages/driver/patches
packages/driver/lib -> /root/cypress/dist/linux/packages/driver/lib
packages/desktop-gui/package.json -> /root/cypress/dist/linux/packages/desktop-gui/package.json
packages/desktop-gui/lib/gui.js -> /root/cypress/dist/linux/packages/desktop-gui/lib/gui.js
packages/desktop-gui/!lib/**/*-spec.jsx -> /root/cypress/dist/linux/packages/desktop-gui/!lib/**/*-spec.jsx
packages/desktop-gui/!lib/**/*-spec.jsx -> /root/cypress/dist/linux/packages/desktop-gui/!lib/**/*-spec.jsx
packages/desktop-gui/!lib/**/*-spec.jsx -> /root/cypress/dist/linux/packages/desktop-gui/!lib/**/*-spec.jsx
packages/desktop-gui/!lib/**/*-spec.jsx -> /root/cypress/dist/linux/packages/desktop-gui/!lib/**/*-spec.jsx
packages/desktop-gui/!lib/**/*-spec.jsx -> /root/cypress/dist/linux/packages/desktop-gui/!lib/**/*-spec.jsx
🔥 deploy error
Error: operation timed out after 4005 ms, 5 tries with error: ENOENT: no such file or directory, stat 'packages/desktop-gui/!lib/**/*-spec.jsx'
    at /root/cypress/node_modules/bluebird-retry/lib/bluebird-retry.js:109:35
    at tryCatcher (/root/cypress/node_modules/bluebird/js/release/util.js:16:23)
```

**Comment by jennifer-shehane** [comment]
Hi, we're changing our contributing workflow to use Draft PRs instead of `[WIP]` in the title. I've converted this PR to a Draft PR and removed `[WIP]` from the title. 

Please [mark the PR as **Ready for Review**](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request#marking-a-pull-request-as-ready-for-review) when you're ready for a Cypress team member to review the PR.

File: packages/desktop-gui/src/settings/settings-spec.jsx [hunk 6]
```diff
@@ -0,0 +1,11 @@
+/// <reference types="cypress" />
+// import Settings from './settings'
+
+// NOTE: cannot import Configuration - regeneratorRuntime undefined error
```
Comment (JessicaSachs) [inline packages/desktop-gui/src/settings/settings-spec.jsx:4]: This is interesting. Can you explain it? This is babel-related right?

Comment (bahmutov) [reply packages/desktop-gui/src/settings/settings-spec.jsx:4]: yeah, it is some strange async / await stuff. I need to update this example, we can actually load this

File: packages/desktop-gui/cypress-component.json [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+  "pluginsFile": "../../node_modules/cypress-react-unit-test/plugins/load-webpack",
+  "env": {
+    "webpackFilename": "webpack.config.ts",
+    "coverage": false
+  },
+  "testFiles": "**/*-spec.{js,jsx}",
+  "experimentalComponentTesting": true,
+  "componentFolder": "src",
+  "nodeVersion": "system"
+}
```
Comment (brian-mann) [inline packages/desktop-gui/cypress-component.json:14]: it doesn't feel right that we need to create a new `cypress-component.json` file. ideally this should be baked into the existing `cypress.json`. Our desktop gui tests are pretty vanilla.

File: packages/desktop-gui/cypress/fixtures/component_specs.json [hunk 8]
```diff
@@ -0,0 +1,25 @@
+{
+  "integration": [
```
Comment (brian-mann) [inline packages/desktop-gui/cypress/fixtures/component_specs.json:2]: i think we should just deprecate `integrationFolder` and rename it to `e2eFolder`, but not make this breaking changes. We could just display a warning in the cypress gui / stdout telling you to update it.

With that said, we could go ahead and rename it to be officially `E2E Tests` and in the url with `cypress/e2e/<spec>`

Comment (bahmutov) [reply packages/desktop-gui/cypress/fixtures/component_specs.json:2]: can it be separate pull request, please

**Comment by bahmutov** [comment]
well, using same webpack as Cypress desktop gui breaks the specs, ughh, exactly why I wanted to separate component cypress.json from regular settings

```
  Running:  settings_spec.js                                                                (1 of 2)

Oops...we found an error preparing this test file:

  cypress/integration/settings_spec.js

The error was:

Error: Webpack Compilation Error
./cypress/integration/settings_spec.js
Module not found: Error: Can't resolve 'lodash/fp' in '/root/cypress/packages/desktop-gui/cypress/integration'
resolve 'lodash/fp' in '/root/cypress/packages/desktop-gui/cypress/integration'
  Parsed request is a module
  using description file: /root/cypress/packages/desktop-gui/package.json (relative path: ./cypress/integration)
    aliased with mapping 'lodash': '/root/cypress/node_modules/lodash/lodash.js' to '/root/cypress/node_modules/lodash/lodash.js/fp'
      using description file: /root/cypress/packages/desktop-gui/package.json (relative path: ./cypress/integration)
        Field 'browser' doesn't contain a valid alias configuration
        using description file: /root/cypress/node_modules/lodash/package.json (relative path: ./lodash.js/fp)
          no extension
            Field 'browser' doesn't contain a valid alias configuration
            /root/cypress/node_modules/lodash/lodash.js/fp doesn't exist
          .ts
            Field 'browser' doesn't contain a valid alias configuration
            /root/cypress/node_modules/lodash/lodash.js/fp.ts doesn't exist
```

**Comment by bahmutov** [comment]
Removed separate cypress.json for components testing and resolved problems

File: packages/server/test/integration/http_requests_spec.coffee (hunk 3 above)
Comment (brian-mann) [reply packages/server/test/integration/http_requests_spec.coffee:53]: yeah good

**Pull request merged**